src/MIDebugEngine/Engine.Impl/CygwinFileMapper.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using MICore;
+
+namespace Microsoft.MIDebugEngine
+{
+    class CygwinFilePathMapper
+    {
+        private DebuggedProcess _debuggedProcess;
+        private Dictionary<string, string> _cygwinToWindows;
+
+        public CygwinFilePathMapper(DebuggedProcess debuggedProcess)
+        {
+            this._debuggedProcess = debuggedProcess;
+            this._cygwinToWindows = new Dictionary<string, string>();
+        }
+
+        public string MapCygwinToWindows(string origCygwinPath)
+        {
+            if (!(_debuggedProcess.LaunchOptions is LocalLaunchOptions))
+            {
+                return origCygwinPath;
+            }
+
+            LocalLaunchOptions localLaunchOptions = (LocalLaunchOptions)_debuggedProcess.LaunchOptions;
+
+            string cygwinPath = origCygwinPath.Replace('\\', '/');
+
+            string windowsPath = cygwinPath;
+
+            lock (this._cygwinToWindows)
+            {
+                if (!this._cygwinToWindows.TryGetValue(cygwinPath, out windowsPath))
+                {
+                    if (!LaunchCygPathAndReadResult(cygwinPath, localLaunchOptions.MIDebuggerPath, out windowsPath))
+                    {
+                        return origCygwinPath;
+                    }
+
+                    this._cygwinToWindows.Add(cygwinPath, windowsPath);
+                }
+            }
+
+            return windowsPath;
+        }
+
+        // There is an issue launching Cygwin apps that if a process is launched using a bitness mismatched console,
+        // process launch will fail. To avoid that, launch cygpath with its own console. This requires calling CreateProcess 
+        // directly because the creation flags are not exposed in System.Diagnostics.Process. 
+        //
+        // Return true if successful. False otherwise.
+        private bool LaunchCygPathAndReadResult(string cygwinPath, string miDebuggerPath, out string windowsPath)
```

Comment:
I don't typically use static as an optimization. The savings are very small and opens the code up to being used at inappropriate times. The purpose of static is to allow the method to be called from the class.